File: bot.js
// bot.js
const { Telegraf } = require('telegraf');
const fetch = require('node-fetch');
const Aria2 = require('aria2');

// Load bot token from environment
const BOT_TOKEN = process.env.BOT_TOKEN;

if (!BOT_TOKEN) {
  console.error('❌ BOT_TOKEN not set');
  process.exit(1);
}

const bot = new Telegraf(BOT_TOKEN);

const aria2 = new Aria2({
  host: 'localhost',
  port: 6800,
  secure: false,
  secret: '',
  path: '/jsonrpc'
});

aria2.open().catch(console.error);

async function startAria2Download(url, ctx) {
  try {
    const gid = await aria2.call('addUri', [url]);
    await ctx.reply(`✅ Download started!\nGID: \`${gid}\``, { parse_mode: 'Markdown' });
  } catch (err) {
    console.error('Aria2 Error:', err);
    await ctx.reply(`❌ Failed to start download: ${err.message}`);
  }
}

async function handleTeraBox(link, ctx) {
  try {
    await ctx.reply('🔍 Getting download link...');
    const response = await fetch('https://ytshorts.savetube.me/api/v1/terabox-downloader', {
      method: 'POST',
      body: new URLSearchParams({ url: link }),
      headers: { 'User-Agent': 'Mozilla/5.0' }
    });

    if (!response.ok) throw new Error('API error');
    const data = await response.json();

    const fast = data.response[0].resolutions['Fast Download'];
    const slow = data.response[0].resolutions['HD Video'];

    const testFast = await fetch(fast);
    if (testFast.ok && testFast.headers.get('content-type').includes('application/octet-stream')) {
      await ctx.reply('🚀 Using fast link!');
      await startAria2Download(fast, ctx);
    } else {
      await ctx.reply('⚠️ Fast link failed. Using HD fallback.');
      await startAria2Download(slow, ctx);
    }
  } catch (err) {
    console.error('TeraBox Error:', err);
    await ctx.reply(`❌ Error: ${err.message}`);
  }
}

bot.start((ctx) => ctx.reply('👋 Send me a TeraBox link to start downloading.'));
bot.on('text', async (ctx) => {
  const msg = ctx.message.text;
  if (msg.includes('terabox.com')) {
    await handleTeraBox(msg, ctx);
  } else {
    ctx.reply('❌ Send a valid TeraBox link.');
  }
});

bot.launch().then(() => console.log('🤖 Bot running...'));

File: bot.py
# bot.py

import asyncio
import os
import logging
import aiohttp
import aria2p
from pyrogram import Client, filters

logging.basicConfig(level=logging.INFO)

BOT_TOKEN = os.environ.get("BOT_TOKEN")
API_ID = int(os.environ.get("API_ID", "12345"))
API_HASH = os.environ.get("API_HASH", "your_api_hash")

app = Client("terabox-bot", bot_token=BOT_TOKEN, api_id=API_ID, api_hash=API_HASH)

aria2 = aria2p.API(
    aria2p.Client(host="http://localhost", port=6800, secret="")
)

async def get_terabox_link(link):
    async with aiohttp.ClientSession() as session:
        try:
            response = await session.post(
                "https://ytshorts.savetube.me/api/v1/terabox-downloader",
                data={"url": link},
                headers={"User-Agent": "Mozilla/5.0"}
            )
            data = await response.json()
            fast = data["response"][0]["resolutions"].get("Fast Download")
            slow = data["response"][0]["resolutions"].get("HD Video")
            return fast or slow
        except Exception as e:
            logging.error(f"Error getting link: {e}")
            return None

@app.on_message(filters.command("start"))
async def start(client, message):
    await message.reply("👋 Send me a TeraBox link to download.")

@app.on_message(filters.text & ~filters.command("start"))
async def handle_link(client, message):
    url = message.text.strip()
    await message.reply("🔍 Generating download link...")
    download_url = await get_terabox_link(url)
    if not download_url:
        await message.reply("❌ Failed to generate download link.")
        return

    await message.reply("📥 Starting download with Aria2...")
    try:
        download = aria2.add_uris([download_url])
        await message.reply(f"✅ Download started!\n\nFile: `{download.name}`", parse_mode="markdown")
    except Exception as e:
        await message.reply(f"❌ Aria2 error: {e}")

if __name__ == "__main__":
    app.run()
